### PR TITLE
[s2s trainer] fix DP mode

### PR DESCRIPTION
**Pull request by stas00**
This PR:

* [x] fixes https://github.com/huggingface/transformers/issues/8822 which currently crashes under multigpu and w/o an explicit ddp mode
* [x] adds tests
* [x] makes `finetune_trainer.py` executable/runnable

@patrickvonplaten, @sgugger

### CODE REVIEW

**Review by patrickvonplaten — APPROVED** [review]
Very clean! Thanks

File: examples/seq2seq/seq2seq_trainer.py [hunk 1]
```diff
@@ -122,7 +122,8 @@ def _get_train_sampler(self) -> Optional[torch.utils.data.sampler.Sampler]:
         else:
             if self.args.sortish_sampler:
                 self.train_dataset.make_sortish_sampler(
-                    self.args.per_device_train_batch_size, distributed=self.args.n_gpu > 1
+                    self.args.per_device_train_batch_size,
+                    distributed=(self.args.n_gpu > 1 and self.args.local_rank != -1),
```
Comment (sgugger) [inline examples/seq2seq/seq2seq_trainer.py:126]: This is always `False` because of [this line](https://github.com/huggingface/transformers/blob/19fa01ce2acdffab84014fa99df2f107b67059ce/src/transformers/training_args.py#L431). The `n_gpu` property is (as documented) the number of GPUs used by the given process. So:
- for `DataParallel`, `local_rank==-1` and `n_gpu` is the number of gpus of the machine used
- for `DistributedDataParallel`, `local_rank` is an integer between 0 and the number of GPUs of the machine - 1, and `n_gpu=1` (each process uses one GPU).

is it possible you wanted to put an `or` here?

File: examples/seq2seq/test_finetune_trainer.py [hunk 2]
```diff
@@ -195,12 +218,17 @@ def run_trainer(self, eval_steps: int, max_len: str, model_name: str, num_train_
...
-            cmd = [sys.executable] + distributed_args + args
+            if dist is None:
+                assert f"for n_gpu={n_gpu} pass `dp` or `dpp` for the `dist` arg"
+            launch_args = []
+            if dist == "ddp":
+                launch_args = f"""
+                    -m torch.distributed.launch
+                    --nproc_per_node={n_gpu}
+                """.split()
+            launch_args.append(f"{self.test_file_dir}/finetune_trainer.py")
```
Comment (sgugger) [inline examples/seq2seq/test_finetune_trainer.py:229]: If `dist=dp` the `n_gpu` argument is not used here, so maybe this should be in a separate test?

Comment (stas00) [reply examples/seq2seq/test_finetune_trainer.py:229]: Sorry, I don't follow your question - could you please expand on it?

Currently `dist=dp` runs here, it is just `dist=dpp` is a special case where it pushes extra args and hence needs `--nproc_per_node={n_gpu}`

Under `dist=dp` the `n_gpu` is sorted out inside trainer, is it not? or do you suggest it should have been passed as an argument to the trainer?

File: examples/seq2seq/seq2seq_trainer.py (hunk 1 above)
Comment (stas00) [reply examples/seq2seq/seq2seq_trainer.py:126]: Oh, I see - thank you for catching that. So I didn't solve the actual problem, but had a lack of hiding it under the carpet.

The problem is that the `distributed=...` is wrong here - it is coded to expect ddp and not dp. dp doesn't have world_size/etc and so it fails, so should it be called `dpp` instead?
```
                self.train_dataset.make_sortish_sampler(
                    self.args.per_device_train_batch_size, dpp=self.args.local_rank != -1)
```
or re-coded to handle dp too? I don't know the initial intention.

In either case, we are back to the request at https://github.com/huggingface/transformers/issues/8822 - let's have a simple flag so that the downstream know which mode is it under and not via checking ranks and n_gpus which is very confusing.

**Comment by stas00** [comment]
moving the discussion out of the review commentary as it disappears as soon as it's resolved, so it's best to discuss it in the normal comments.

-------------

Oh, I see - thank you for catching that. So I didn't solve the actual problem, but had a luck of hiding it under the carpet.

The problem is that the `distributed=...` is wrong here - it is currently coded to expect ddp when `distributed==True` and not dp. dp doesn't have world_size/etc and so it fails, so should it be called `dpp` instead of distributed? But in any case the correct solution is then:
```
                self.train_dataset.make_sortish_sampler(
                    self.args.per_device_train_batch_size, distributed=self.args.local_rank != -1)
```
or re-coded to handle dp too? I don't know the initial intention - should it support `sortish_sampler` under dp or not?

In either case, we are back to the request at https://github.com/huggingface/transformers/issues/8822 - let's have a simple flag so that the downstream code knows which mode it is under and not via checking ranks and n_gpus which is very confusing.

**Comment by stas00** [comment]
Here is where the problem happens with dp:

https://github.com/huggingface/transformers/blob/9995a341c9d68a9963d86c506d17330b3ad813f9/examples/seq2seq/utils.py#L361-L373

So `dist.is_available()` returns True under `dp`, but `dist.get_world_size()` fails, since it only works under dpp and requires `torch.distributed.init_process_group()` which doesn't get called under dpp.

File: examples/seq2seq/test_finetune_trainer.py (hunk 2 above)
Comment (sgugger) [reply examples/seq2seq/test_finetune_trainer.py:229]: Using `dist='dp'` does not require the `n_gpu` argument, yet we only launch it inside a test regarding that `n_gpu`, which I find weird.
I'm not familiar with this part of the code so maybe I'm missing something here.

**Comment by sgugger** [comment]
In `DataParallel` mode, you don't need to do anything to your datalaoder (only in DistributedDataParallel where you need to split the batches across the various processes somehow) so you should make a regular datalaoder in that case.
In general, the only proper way to detect if you are in distributed data parallel is to look at the test `local_rank != -1` as `torch.distributed` can give you false information there. I agree it would all be much easier if the training arguments contained something that directly gives the distributed environment.

File: examples/seq2seq/test_finetune_trainer.py (hunk 2 above)
Comment (stas00) [reply examples/seq2seq/test_finetune_trainer.py:229]: since this is wrapper able to run under any setup,  it uses `n_gpu` to know whether to:

1. launch a distributed test (dp or dpp)
2. pass `--nproc_per_node={n_gpu}` in case of `dpp`

We could explicitly pass `n_gpu` in case of `dist=dp` if it makes things more clear.

Also perhaps it's not clear how `dist=dp` fits into this picture so may be the following would be a more readable code:
```
        n_gpu = get_gpu_count()
        if n_gpu > 1:
            if dist == "dp":
                launch_args = []
            elif dist == "ddp":
                launch_args = f"""
                    -m torch.distributed.launch
                    --nproc_per_node={n_gpu}
                """.split()
            else:
                assert f"for n_gpu={n_gpu} pass `dp` or `dpp` for the `dist` arg"
            launch_args.append(f"{self.test_file_dir}/finetune_trainer.py")
```
```

**Comment by stas00** [comment]
> In `DataParallel` mode, you don't need to do anything to your datalaoder (only in DistributedDataParallel where you need to split the batches across the various processes somehow) so you should make a regular datalaoder in that case.

Great, so then should we change the signature to make it clear ddp is wanted and not any distributed:

```
- def make_sortish_sampler(self, batch_size, distributed=False, shuffle=True, **kwargs):
+ def make_sortish_sampler(self, batch_size, ddp=False, shuffle=True, **kwargs):
```

and adjust the invocations accordingly?

> In general, the only proper way to detect if you are in distributed data parallel is to look at the test `local_rank != -1` as `torch.distributed` can give you false information there. I agree it would all be much easier if the training arguments contained something that directly gives the distributed environment.

Great. Should we create a feature request for that?

**Comment by sgugger** [comment]
I think there is a misunderstanding on the terminology: `DataParallel` is not distributed: distributed means launching several processes with the same script. The package `torch.distributed` does not return anything useful for `DataParallel` and ddp stands for *distributed* data parallel, so leaving that argument as distributed seems better to me.

> Great. Should we create a feature request for that?

We can do that, yes.

File: examples/seq2seq/test_finetune_trainer.py (hunk 2 above)
Comment (sgugger) [reply examples/seq2seq/test_finetune_trainer.py:229]: I guess I don't understand why you need to make a special case for dp compared to the regular script. Launching a script in dp is the same as launching it period. Pytest will use dp if you have several GPUs and the tests you are running use it (always have to set a `CUDA_VISIBLE_DEVICES="1"` on my machine to avoid that for instance).

**Comment by stas00** [comment]
If you stick to the specific implementation, yes, dpp is the only distributed. But logically it doesn't make sense. DP is just as distributed as DPP, just isn't using the `torch.distributed`, so it's not a very clear distinction and will lead to such confusions all over. 

As an example if you look at this function usage pattern it's mostly `dataset.make_sortish_sampler(batch_size, distributed=self.hparams.gpus > 1)` which clearly implies for any multi gpu mode (and erroneously so).

**Comment by sgugger** [comment]
I disagree, in the sense that code use PyTorch should stick with the PyTorch naming conventions. They chose to have a not distributed `DataParallel`, so we should honor that in our naming as well. In Distributed data parallel, you have to use a `DistributedSampler` (but not in `DataParallel`) etc. Those are all *parallel* modes (as you're training with multiple GPUs) but only one is *distributed*.

**Comment by stas00** [comment]
That is a reasonable choice to follow. I'm only flagging how this leads to coding errors when a developer assumes that n_gpu> 1 == ddp. So perhaps some extra support is needed there.

File: examples/seq2seq/test_finetune_trainer.py (hunk 2 above)
Comment (stas00) [reply examples/seq2seq/test_finetune_trainer.py:229]: Ah, I finally see what you mean - thank you for clarifying, @sgugger. Let me experiment and get back to you.

**Comment by sgugger** [comment]
Let's see how it goes once we add the "distributed_env" to `TrainingArguments`!

File: examples/seq2seq/test_finetune_trainer.py (hunk 2 above)
Comment (stas00) [reply examples/seq2seq/test_finetune_trainer.py]: OK, sorted it out - simplified and re-tested. Thank you for that dialog, @sgugger - a much better outcome thanks to you.

File: examples/seq2seq/seq2seq_trainer.py (hunk 1 above)
Comment (stas00) [reply examples/seq2seq/seq2seq_trainer.py]: This got resolved with using: `distributed=(self.args.local_rank != -1),`

**Comment by stas00** [comment]
@sgugger, please kindly review at your convenience - all should be good - CI failures are unrelated. Thank you!

**Review by sgugger — APPROVED** [review]
Perfect, thanks a lot for humoring me and my annoying comments :-)

**Comment by stas00** [comment]
> Perfect, thanks a lot for humoring me and my annoying comments :-)

On the contrary, your comments were excellent and to the point. 

I was just slow on getting your point of view since in my mind if we solve a problem on multiple gpus it's distributed across multiple-gpus, regardless of the way it's implemented. But here distributed means distributed across multiple processes. Different semantics.

**Comment by stas00** [comment]
So this is probably wrong too:
```
examples/seq2seq/finetune.py:            sampler = dataset.make_sortish_sampler(batch_size, distributed=self.hparams.gpus > 1)
```

**Comment by sgugger** [comment]
Dunno how PL works.

**Comment by stas00** [comment]
> Let's see how it goes once we add the "distributed_env" to `TrainingArguments`!

Added a feature request: https://github.com/huggingface/transformers/issues/8858

**Comment by rabeehk** [comment]
Thank you HuggingFace Team and @stas00 , I cannot express how much I appreciate your efforts.
